src/monitor/group_state_machine.c

```diff
@@ -570,6 +783,506 @@ ProceedGroupStateForPrimaryNode(AutoFailoverNode *primaryNode)
+							   AutoFailoverNode *primaryNode)
+{
+	List *standbyNodesGroupList =
+		primaryNode == NULL
+		? AutoFailoverNodeGroup(activeNode->formationId, activeNode->groupId)
+		: AutoFailoverOtherNodesList(primaryNode);
+	CandidateList candidateList = { 0 };
+
+	/*
+	 * Done with the single standby code path, now we have several standby
+	 * nodes that might all be candidate for failover, or just some of them.
+	 *
+	 * The first order of business though is to determine if a failover is
+	 * currently happening, by looping over all the nodes in case one of them
+	 * has already been selected as the failover candidate.
+	 */
+	AutoFailoverNode *candidateNode =
+		FindCandidateNodeBeingPromoted(standbyNodesGroupList);
+
+	/*
+	 * If a failover is in progress, continue driving it.
+	 */
+	if (candidateNode != NULL)
+	{
+		elog(LOG, "Found candidate node %d (%s:%d)",
+			 candidateNode->nodeId,
+			 candidateNode->nodeHost,
+			 candidateNode->nodePort);
+
+		return ProceedWithMSFailover(activeNode, candidateNode);
+	}
+
+	/*
+	 * Now, have all our candidates for failover report the most recent LSN
+	 * they managed to receive. We build the list of nodes that we consider as
+	 * failover candidates into candidateNodesGroupList.
+	 *
+	 * When every one of the nodes in that list has reported its LSN position,
+	 * then we select a node from the just built candidateNodesGroupList to
+	 * promote.
+	 *
+	 * It might well be that in this call to node_active() only a part of the
+	 * candidates have reported their LSN position yet. Then we refrain from
+	 * selecting any in this round, expecting a future call to node_active() to
+	 * be the kicker.
+	 *
+	 * This design also allows for nodes to concurrently be put to maintenance
+	 * or get unhealthy: then the next call to node_active() might build a
+	 * different candidateNodesGroupList in which every node has reported their
+	 * LSN position, allowing progress to be made.
+	 */
+	BuildCandidateList(standbyNodesGroupList, &candidateList);
+
+	/*
+	 * Time to select a candidate?
+	 *
+	 * We reach this code when we don't have an healthy primary anymore, it's
+	 * been demoted or is draining now. Most probably it's dead.
+	 *
+	 * Before we enter the selection process, we must have collected the last
+	 * received LSN from ALL the standby nodes that are considered as a
+	 * candidate (thanks to the FSM transition secondary -> report_lsn), and
+	 * now we need to select one of the failover candidates.
+	 *
+	 * We might fail to have reportedLSNCount == candidateCount in the
+	 * following cases:
+	 *
+	 * - the activeNode is the first to report its LSN, we didn't hear from the
+	 *   other nodes yet; in that case we want to wait until we hear from them,
+	 *   it is expected to happen within the next 5 seconds.
+	 *
+	 * - a standby that was counted as a candidate is not reporting; most
+	 *   probably it's not healthy anymore (or has been put to maintenance
```

Comment:
As  talked in person: I guess the second part of the comment doesn't exactly reflect the behaviour: `and the next call to node_active() should be able to account for that new global state and make progress: the now faulty standby will NOT be counted as candidate anymore.`

If a node is unhealthy, we'll NOT make progress because we might loose data.  We should probably reflect it in the comment?

Reply:
The comment has been updated to the following now:

``` c
	 * We might have a non-zero missingNodesCount in the following cases:
	 *
	 * - the activeNode is the first to report its LSN, we didn't hear from the
	 *   other nodes yet; in that case we want to wait until we hear from them,
	 *   it is expected to happen within the next seconds.
	 *
	 * - a standby that was counted as a candidate is not reporting; most
	 *   probably it's not healthy anymore. In that case we might not be able
	 *   to proceed with the failover without a risk of data-loss. We avoid
	 *   data-loss situations by waiting until the secondary is back.
```

Having written that, I'm not too happy about the fact that we block the group state machine when we have unhealthy standby nodes. I think we should check if we actually need them, which is the case if we loose the candidate node, or the most advanced node, or both, but that's it, other ones should be allowed to be unhealthy.

Reply:
Ok I didn't like that too much so I have reviewed entirely this part of the code, to make it more resilient and robust to some situations. It should be better already, but unfortunately your new test is now failing. Because I'm not familiar with your test, I though I would still push my changes and we'll have a look tomorrow on how to best fix the code again so that your test pass again.